sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,228 @@
+#title          :discretization.py
+#description    :Implementation of a few discretization algorithms
+#author         :Henry Lin <hlin117@gmail.com>
+#==============================================================================
+
+from __future__ import division
+from math import log
+import numpy as np
+from scipy.stats import entropy
+from sklearn.base import BaseEstimator
+
+class MDLP(BaseEstimator):
+    """Implements the MDLP discretization criterion from Usama Fayyad's
+    paper "Multi-Interval Discretization of Continuous-Valued Attributes
+    for Classification Learning"
+    """
+
+    def __init__(self, **params):
+        self.set_params(**params)
+
+    def get_params(self, deep=True):
+        return {"continuous_columns": self.continuous_columns,
+                "min_depth": self.min_depth}
+
+    def set_params(self, **params):
+        self.continuous_columns = params.get("continous_columns")
+        self.min_depth = params.get("min_depth", 0)
+        return self
+
+    def fit(self, X, y):
+        """Finds the intervals of interest from the input data.
+        """
+        if type(X) is list:
+            X = np.array(X)
+        if type(y) is list:
+            y = np.array(y)
+        if self.continuous_columns is None:
+            self.continuous_columns = range(X.shape[1])
+        if(len(X.shape) != 2):
+            raise ValueError("MDLP can ony be applied to input ndarrays of "
+                             "size 2.")
+        if(X.shape[0] != y.shape[0]):
+            raise ValueError("Number of samples in X does not match number "
+                             "of targets in y")
+
+        self.num_classes_ = set(y)
+        self.intervals_ = dict()
+
+        for index, col in enumerate(X.T):
+            if index not in self.continuous_columns:
+                continue
+            intervals = self._mdlp(col, y)
+            intervals.sort(key=lambda interval: interval[0])
+            self.intervals_[index] = [(interval, i)
+                                     for i, interval in enumerate(intervals)]
+        return self
+
+    def transform(self, X):
```

Comment:
I've not looked at how you've stored the model, etc., but I would have expected a discretiser's `transform` method to work with a sorted array of boundaries per feature, such as `boundaries=array([-inf, 0, 2, 5, 10, inf])` (the infs may not be necessary) then employ `np.searchsorted`, e.g. `np.searchsorted(boundaries, [6, 3, 100, 0, 1])` returns `array([4, 3, 5, 1, 2])`.

So `transform` should be something like:
```python
out = X.copy()
for i, boundaries in zip(self.continuous_columns, self.boundaries_):
    out[:, i] = np.searchsorted(boundaries, X[:, i])
return out
```

Please also accept an unused `y` argument (`y=None`) for API consistency.